### ISSUE TITLE
Gamepad-specific DOM keys / location?

### ISSUE THREAD

**Issue by RByers** [w3c/uievents#79]
This started as a [blink implementation discussion](https://codereview.chromium.org/1680143002/), but I think the important question here is really a platform API decision.

If a (otherwise normal) web-page wants to optimize it's behavior when being navigated by a gamepad, is the [Gamepad API](https://developer.mozilla.org/en-US/docs/Web/API/Gamepad_API/Using_the_Gamepad_API) always the best choice?  Eg. when the Gamepad API is in use, gamepad browser navigation features (like built-in scrolling controls) must be disabled, and it's probably not easy / practical for the web developer to re-implement scrolling support (matching the browser default conventions) on top of the low-level Gamepad API.  Also it's weird for a normal document-style page to need a "game loop" to detect (probably rare) button presses, rather than just having events the way a keyboard does.

So maybe it's also reasonable (as @garykac [suggests](https://codereview.chromium.org/1680143002/#msg29)) to generate Keyboard events for gamepad buttons?

@sgraham @garykac Thoughts?

**Comment by garykac** [comment]
The only reason to consider this is if current browsers already implement this behavior. In that case, we need to spec it for compatibility (or file bugs to have the behavior removed).

As far as I can tell, no browser currently generate keyboard events for gamepads (although there are 3rd-party extensions that can do this).

Given that no browser currently ships this feature, there is no reason to consider adding it.

**Comment by alexelias** [comment]
As I mentioned on the code review, I don't think we should be adding any default useragent behavior that cannot be detected and preventDefault()ed by web authors.  The problem is that there's no sane way to preventDefault browser behaviors with the poll-based gamepad API -- it really needs an event-based model.  (I guess this is also a problem with the gamepad scrolling behavior we've already allowed into Chrome, which we might want to revisit.)

The most natural solution would be to simply forward the artificial gamepad keyboard events which Android framework is already supplying to Chromium.  In general I think it's a reasonable philosophy to forward to JS without information loss all input events that are provided to us by the OS, even if this reveals certain platform differences.  (As long as we're only *revealing* them rather than introducing them.)  That's usually preferable to dropping useful information on the floor even if that would create an artificial cross-platform consistency.

**Comment by RByers** [comment]
I agree that we should be looking at this bigger than just what browsers do today (which is to not support the gamepad navigation use case except via hacks).  The GamePad API makes sense for the gaming scenario, but for browsing with a GamePad, it's really just another input device which should be supported by the existing input APIs.

To clarify the concrete use case:
 - User agent wants to connect "button A" to certain default actions like navigating a link and clicking a checkbox
 - Developer should be able to override that behavior somehow if they want to customize the gamepad-driven experience
 - Button_A shouldn't necessarily have the exact same semantics as any existing keyboard event.  Eg. it shouldn't add any whitespace to a text field.

We've long handled more unusual input devices by just pretending they're something they're not.  Sometimes that's necessary for web compat (eg. touch screens generating mouse events in limited scenarios).  But subject to compat constraints I think in general we should strive to expose all the information available to the native platform.  I mean, if [TVSatelliteToggle](https://w3c.github.io/uievents-key/#keys-tv) isn't too niche to get a key value, then surely `GamepadButtonA` isn't either.

[Here's some details of the Android API](http://developer.android.com/training/game-controllers/controller-input.html#button) we're trying to expose here.  @alexelias any idea what the API on other OSes looks like?  If we're going to specify something here, there's the practical matter of figuring out how exactly to assign [code](https://w3c.github.io/uievents-code/) and [key](https://w3c.github.io/uievents-key/) values.

**Comment by alexelias** [comment]
To the best of my knowledge, other OSes all provide a poll-based API for gamepad; Android is the only OS to issue events for it instead.

One other important detail is that the stick axis scrolling is a harder problem.  If the user snaps the stick to one extreme and then keeps it there, we only get one motion event, and then Chromium has to create a self-driven animation to scroll.  So something like the "beforescroll" event model would be a better fit if we want to provide fine-grained control over joystick scroll.  That said, coarse-grained disabling using "overflow: hidden" may be sufficient in practice anyway.

**Comment by RByers** [comment]
Here's a specific minimal proposal to try to unblock the [chromium issue](https://codereview.chromium.org/1680143002/):

- Add `DOM_KEY_LOCATION_GAMEPAD` that indicates the key event comes from a device that's also exposed by the [W3C Gamepad API](https://w3c.github.io/gamepad/)
- Add (at least) `Button0` to the [code spec](https://w3c.github.io/uievents-code/), defined to be equivalent to `buttons[0]` in [the standard Gamepad API mapping](https://w3c.github.io/gamepad/#remapping)
- Add (at least) `ButtonA` to the [key spec](https://w3c.github.io/uievents-key/).  
  - Alternately perhaps it's not unreasonable to re-use "Execute" or "Navigate" when chromium wants to map BUTTON_A to a general execution action?
